vm/src/bytesinner.rs

```diff
@@ -80,62 +78,114 @@ impl ByteInnerNewOptions {
         Ok(vec![0; size].into())
     }
 
-    fn check_args(self, vm: &VirtualMachine) -> PyResult<()> {
-        if let OptionalArg::Present(_) = self.encoding {
-            Err(vm.new_type_error("encoding without a string argument".to_owned()))
-        } else if let OptionalArg::Present(_) = self.errors {
-            Err(vm.new_type_error("errors without a string argument".to_owned()))
-        } else {
-            Ok(())
-        }
+    fn string_without_encoding(vm: &VirtualMachine) -> PyBaseExceptionRef {
+        vm.new_type_error("string argument without an encoding".to_owned())
     }
 
-    pub fn get_bytes(mut self, cls: PyTypeRef, vm: &VirtualMachine) -> PyResult<PyBytesRef> {
-        let inner = if let OptionalArg::Present(source) = self.source.take() {
-            if cls.is(&PyBytes::class(vm)) {
-                if let Ok(source) = source.clone().downcast_exact(vm) {
-                    return self.check_args(vm).map(|()| source);
-                }
-            }
+    fn encoding_without_string(vm: &VirtualMachine) -> PyBaseExceptionRef {
+        vm.new_type_error("encoding without a string argument".to_owned())
+    }
 
-            match_class!(match source {
-                s @ PyStr => Self::get_value_from_string(s, self.encoding, self.errors, vm),
-                i @ PyInt => {
-                    self.check_args(vm)?;
-                    Self::get_value_from_size(i, vm)
-                }
-                obj => {
-                    self.check_args(vm)?;
-                    if let Some(bytes_method) = vm.get_method(obj.clone(), "__bytes__") {
-                        let bytes = vm.invoke(&bytes_method?, ())?;
-                        PyBytesInner::try_from_borrowed_object(vm, &bytes)
-                    } else {
-                        Self::get_value_from_source(obj, vm)
+    fn errors_without_string(vm: &VirtualMachine) -> PyBaseExceptionRef {
+        vm.new_type_error("errors without a string argument".to_owned())
+    }
+
+    pub fn get_bytes(self, cls: PyTypeRef, vm: &VirtualMachine) -> PyResult<PyBytesRef> {
+        match (self.source, self.encoding, self.errors) {
+            (OptionalArg::Present(obj), OptionalArg::Missing, OptionalArg::Missing) => {
+                // construct an exact bytes from an exact bytes do not clone
+                let obj = if cls.is(PyBytes::class(vm)) {
+                    match obj.downcast_exact::<PyBytes>(vm) {
+                        Ok(b) => {
+                            return Ok(b);
+                        }
+                        Err(obj) => obj,
                     }
+                } else {
+                    obj
+                };
+
+                let inner = if let Some(bytes_method) = vm.get_method(obj.clone(), "__bytes__") {
```

Comment:
any way to check the method existing without clone?

Reply:
don't think so.